v3_async_wip/v3_async_wip/iothub_session.py

```diff
@@ -0,0 +1,283 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import asyncio
+import contextlib
+import ssl
+from typing import Optional, Union, AsyncGenerator, Type
+from types import TracebackType
+
+from v3_async_wip import signing_mechanism as sm
+from v3_async_wip import connection_string as cs
+from v3_async_wip import sastoken as st
+from v3_async_wip import config, models, custom_typing
+from v3_async_wip import iothub_mqtt_client as mqtt
+
+
+class IoTHubSession:
+    def __init__(
+        self,
+        *,
+        hostname: str,  # iothub_hostname?
+        device_id: str,
+        module_id: Optional[str] = None,
+        ssl_context: Optional[ssl.SSLContext] = None,
+        shared_access_key: Optional[str] = None,
+        sastoken_fn: Optional[custom_typing.FunctionOrCoroutine] = None,
+        **kwargs,
+    ) -> None:
+        """
+        :param str device_id: The device identity for the IoT Hub device containing the
+            IoT Hub module
+        :param str module_id: The module identity for the IoT Hub module
+        :param str hostname: Hostname of the IoT Hub or IoT Edge the device should connect to
+        :param ssl_context: Custom SSL context to be used by the client
+            If not provided, a default one will be used
+        :type ssl_context: :class:`ssl.SSLContext`
+        :param str shared_access_key: A key that can be used to generate SAS Tokens
+        :param sastoken_fn: A function or coroutine function that takes no arguments and returns
+            a SAS token string when invoked
+
+        :raises: ValueError if none of 'ssl_context', 'symmetric_key' or 'sastoken_fn' are provided
+        :raises: ValueError if both 'symmetric_key' and 'sastoken_fn' are provided
+        :raises: ValueError if an invalid 'symmetric_key' is provided
+        :raises: TypeError if an invalid keyword argument is provided
+        """
+        # Validate parameters
+        _validate_kwargs(**kwargs)
+        if shared_access_key and sastoken_fn:
+            raise ValueError(
+                "Incompatible authentication - cannot provide both 'shared_access_key' and 'sastoken_fn'"
+            )
+        if not shared_access_key and not sastoken_fn and not ssl_context:
+            raise ValueError(
+                "Missing authentication - must provide one of 'shared_access_key', 'sastoken_fn' or 'ssl_context'"
+            )
+
+        # Set up SAS auth (if using)
+        generator: Optional[st.SasTokenGenerator]
+        # NOTE: Need to keep a reference to the SasTokenProvider so we can stop it during cleanup
+        self._sastoken_provider: Optional[st.SasTokenProvider]
+        if shared_access_key:
+            uri = _format_sas_uri(hostname=hostname, device_id=device_id, module_id=module_id)
```

Comment:
>hostname=hostname [](http://example.com/codeflow?start=34&length=17)

This is wrong because you're overloading the `hostname` variable to be either the `hostname` or the `gatewayhostname` from the connection string. If connecting to a gateway, the sas token gets built using `hostname` and the TLS socket is opened using `gatewayhostname`

Reply:
I will double check this again, but I believe this does work - Ryan and I discussed it with the Edge folks while you were out of office.